tao_compiler/mlir/disc/transforms/disc_compare_simplifier.cc

```diff
@@ -0,0 +1,363 @@
+// Copyright 2022 The BladeDISC Authors. All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Debug.h"
+#include "mlir-hlo/Analysis/shape_component_analysis.h"
+#include "mlir-hlo/Dialect/mhlo/IR/hlo_ops.h"
+#include "mlir/Dialect/Shape/IR/Shape.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+#include "tensorflow/compiler/mlir/disc/transforms/PassDetail.h"
+#include "tensorflow/compiler/mlir/disc/transforms/shape_utils.h"
+
+namespace mlir {
+namespace disc_ral {
+namespace {
+
+// This pattern analysis the operands of CompIOp and try to infer the result
+// at compile time with affine-expression analysis.
+struct ArithComISimplifier : public OpRewritePattern<arith::CmpIOp> {
+  explicit ArithComISimplifier(MLIRContext* context)
+      : OpRewritePattern(context) {}
+
+  LogicalResult matchAndRewrite(arith::CmpIOp op,
+                                PatternRewriter& rewriter) const override {
+    Location loc = op.getLoc();
+    Value lhs = op.getLhs();
+    Value rhs = op.getRhs();
+    auto pred = op.getPredicate();
+
+    // TODO: we plan to replace `ShapeComponentAnalysis` with our implemented
+    // `ShapeAnalysisV2` in the near further.
+    ShapeComponentAnalysis shapeComponentAnalysis;
+    auto lhsInfo = shapeComponentAnalysis.GetValueInfo(lhs);
+    auto rhsInfo = shapeComponentAnalysis.GetValueInfo(rhs);
+    if (!lhsInfo || ((*lhsInfo).size() != 1) || !rhsInfo ||
+        ((*rhsInfo).size() != 1)) {
+      return failure();
+    }
+    auto predict = compareSymbolicExpr((*lhsInfo)[0], (*rhsInfo)[0], pred);
+    if (predict.hasValue()) {
+      Value new_cmpi = rewriter.create<arith::ConstantIntOp>(
+          loc, *predict, rewriter.getIntegerType(1));
+      rewriter.replaceOp(op, new_cmpi);
+      return success();
+    }
+
+    return failure();
+  }
+
+  struct SignState {
```

Comment:
is this necessarily to be a struct, or just enum is good enough?

Reply:
It could be a struct wrapping an enum. We need the helper functions.

Reply:
Done.